### DISCUSSION TITLE
[BE101] PHP 引入頁面的變數問題/GET VS POST的使用問題 [腦洞文1.2&1.3]

### DISCUSSION THREAD

**Discussion by IamTomIamQQ** [Lidemy/forum#129] [其他問題] [answered]
> 　　之前寫到自製TOKEN時，寫到登入後必須讓index.php只有已登入使用者看到部分版面。發生了一些問題(我的腦洞疑問)，想請教一下先賢們：
> 
> ### 問題１
> 　　針對require_once("conn.php");，如此連結了$conn，但當下是否 **$username** 這個變數也被當成全域變數 **引入了？** 所以此時必須設$username =NULL;。（因為寫到一半echo檢查時老是發現我的username是'huli'）
> 
> 　　另外因為$username剛好也要被當作驗證是否有登入，若尚未登入時；由TOKEN讀取此變數值，按理說這會被當成一個 **問題** _(非法?)_ 變數呢，所以開頭才需要設$username =NULL;？  **請問我的推測是正確的嗎?**  
> (我會有這樣的疑問是因為 當下我跑去把$conn中變數改名了，然後在未登入的情況下就跑出問題警告文字)

以上已解答在下疑惑，感謝回答！
-----------------------------------------------------------------------------------------------------------------------
### 問題２(6/7修正問題內容)  
　　抱歉當時急著發問，竟然沒發現自己抄錯段文字了；已修正與補述如下：  

　　當我撰寫index.php大致刻完版面時，並沒有先寫comments引入等等，而是忙著寫登入登出與cookie，因為我想既然SESSION 或token我不太熟就乾脆從這個方向下手好了！(O.O!!!)  
(順著思路亂飄就變成關關難過>忄光然>關關難過>忄光然 ~~永無止境)這是我自己順一次後的感想  

###### 通常登出最簡易的流程大概是
> 1. index中`<a href="./handle_logout.php" class="board__btn">我要登出</a>`
> 2. 使用者點擊，跳到`handle_logout`。
>   　**這裡就是我腦洞開的思路點！**  
>   因為我在想會不會需要驗證發出登出請求地的就是本人，所以才有以下連串怪異寫法  
> 3. handle_logout會刪去cookie然後同時刪去對應的token各個資料欄位
> 

###### 所以index上方登出連結我嘗試改寫成
```
<form class="board__btn" metod="POST" action="./handle_logout.php">
　<input type="hidden" name="logout" value="<?php echo $token;  ?>" />
　　<input type="submit" value="登出" />
</form>'
```
###### 為了測試我在handle_logout加上
```
<?php
if(!empty($_POST['logout'])) {
　echo $_POST['logout'];
}
```
結果得到如下兩個結果當場就嚇矇我了：
handle_logout網址列：  
`http://localhost/board3.0test/handle_logout.php?logout=KEXBXMWHQUUTDWFW`。
必須改成才會得到傳送的TOKEN值(也檢查Devtool/network了確實是GET)
```
if(!empty($_GET['logout'])) {
　echo $_GET['logout'];
}
```

#### 今日嘗試重現當時寫法，再三檢查確認得到的就是GET傳送，但我明明使用POST方法傳遞，為什麼會變成GET呢？冏~

**Comment by aszx87410** [comment]
1. 你 `require_once("conn.php")` 的話就是  conn.php 的內容都會被引入
2. 你的 action 是 `myindex.php`，不是登出的頁面，可以用 devtool 看一下到底送出去什麼 request

**Reply by IamTomIamQQ** [reply]
問題2 我重新檢查然後又發問了(抱歉當時思路亂，筆誤)麻煩幫我看看




PS 上課過程我老是有種錯覺 ，POST就是比GET安全了點，所以我什麼鬼東西 都很想改成POST來寫，然後常常 卡了又卡， 有些感覺沒那麼危險就只好自己覺得妥協，現在基本的範例自己順一次之後想想好像是自己胡思亂想。 但我其實也沒搞懂。

**Comment by aszx87410** [comment, marked as answer]
程式碼不會騙人，最後是 GET 代表 method 絕對沒設好，那為什麼沒設呢？你把 form html 那一段一個字一個字檢查應該就知道了

**Reply by IamTomIamQQ** [reply]
囧rz  我又拼錯字了
